Add original client design to project (#125)
cyberark/conjur-authn-k8s-client: design/authenticator-client-for-container-platforms.md

A. Compute Cluster - A pool of servers or VMs which run a container engine.
B. Container Platform Engine - Schedules application containers across the Compute Cluster.
C. Application Containers - Individual instances of application code which are scheduled onto the Compute Cluster
   by the Platform Engine
D. Authenticator - A platform application whose function is to verify the identity of Application Containers and issue
   them identity credentials. Runs in the compute cluster.
E. Security Service - A service which provides security methods such as role-based access control, permission checks,
   auditing, and access to secrets. May run inside or outside the compute cluster.

## General design

The Authenticator provides two operations: `login` and `authenticate`. `login` is performed once by each client to
obtain a signed certificate. `authenticate` is performed many times by each client in order to obtain a short-lived
access token. Each access token thus obtained is encrypted by the client's public key.

The `login` method resists identity spoofing by installing the signed certificate onto the client using a Container
Platform Engine API.

> Once the client has obtained its signed certificate, it uses the certificate to authenticate, thereby  obtaining a

Reviewer: The server injects the client certificate before this part is done which should be mentioned.

short-lived access token which encrypted by the client's public key. The encryption ensures that an attacker cannot
use a token without having access to the private key as well (which requires access to the client container's memory).

**Note:** in the original design, we also proposed apply IP restrictions to the short-lived access token. This is not
yet implemented.

## Description

### Login

> The Authenticator performs a function similar to a Certificate Authority: verifying a CSR (certificate signing request)

Reviewer: ```suggestion
The Authenticator performs a function similar to a Certificate Authority: verifying a CSR (certificate signing
request) but through Container Platform Engine APIs
```

and issuing a certificate for it. The client container generates a CSR with the following attributes:

- Subject Name: Application identifier (e.g. host identity)
- Subject Alternative Name: Container identifier (e.g. SPIFFE identity of the form
  `spiffe://cluster.local/namespace/{app namespace}/podname/{app pod name}`)

The corresponding private key is stored in a memory file on the client. The client sends the CSR to the Authenticator

> using the HTTPS method `GET /login`. When the Authenticator receives a CSR, it verifies the CSR attributes with the

Reviewer: Nope: https://github.com/cyberark/conjur-authn-k8s-client/blob/922126ad8d603fc7ce32ac76a2eeee0ef5667205/pkg/authenticator/requests.go#L20

Container Platform Engine. The Authenticator CA then signs the CSR using a CA key which it obtains from a Conjur variable,

> and installs the certificate on the client via side channel. The client stores the certificate in a memory file alongside
> the key. The client can use the certificate and key as client TLS information on subsequent requests to the Authenticator.

Reviewer: ```suggestion
and installs the certificate on the client via APIs in async manner. The client stores the signed cert on file and the key is only stored in memory. The client can use the certificate and key as client TLS information on subsequent requests to the Authenticator.
```

> **Note:** the attributes of the application are different in the actual implementation than in the original design, and

Reviewer: ```suggestion
**Note:** The attributes of the application are different in the actual implementation than in the original design, and
```

we have updated this section to reflect the current implementation. See the document history for info on the original
design.

#### Login flow

**Note:** `login` is typically performed once per Client Container. It may be repeated if the client knows that its
certificate is due to expire.

1. Create an application container
1. Client container generates a certificate private key and stores it in memory
1. Client container generates a CSR

> 1. Client container sends the CSR to the Authenticator `login` method

Reviewer: `fmt.Sprintf("%s/inject_client_cert", authnURL)`

1. Authenticator verifies the Subject Name, Subject Alternative Name(s), and any other attributes of the client request.
1. Authenticator signs the certificate

> 1. Authenticator installs the certificate onto the client using a side-channel (Container Platform Service API)

Reviewer: ```suggestion
1. Authenticator installs the certificate onto the client using a side-channel (Container Platform Service API) and in async thread
```

1. Authenticator responds with “OK” on the original request channel

### Authenticate

The client wants to obtain an access token from the Authenticator. It sends an HTTPS request to `POST /authenticate`.

The Authenticator verifies the client certificate, and verifies the certificate metadata.  The Authenticator issues an
access token and sends the access token back to the client.

**Notes:**
- `authenticate` is performed as often as the Client Container needs. Access tokens issued by `authenticate`

>   have a short lifespan (minutes to hours).

Reviewer: ```suggestion
  have a short lifespan (minutes to hours) but the default duration is 8 minutes.
```

cluster.  

Finally, the Authenticator does not itself store any secrets or other sensitive data aside from its own identity
credential.  

Further guidance on securing the Authenticator will be provided by CyberArk and the Platform vendor.

### Why make the client send a CSR rather than generating the private key and cert in the Authenticator?

It's more secure if the Authenticator never knows the client private keys. It's also better to make the clients
provide entropy rather than making the Authenticator provide the entropy for all the keys.

### If am able to spoof a request IP, can I fool the Authenticator into logging me in me as any arbitrary application?

No. When you login, the Authenticator will install the certificate onto the client container using a side-channel
routed through the orchestration controller. So, you will also need to intercept this request, and you will need
to MITM the SSL between the Authenticator and the orchestration controller.

As long as there is proper validation of the CSR's subject name on login, we don't need to worry about ARP spoofing
as the IP should only be one of two factors considered

> ### If am able to spoof a request IP, can I fool the Authenticator into authenticating me in me as any arbitrary application?

Reviewer: ```suggestion
### If am able to spoof a request IP, can I fool the Authenticator into authenticating me as any arbitrary application?
```

Reviewer: This whole bottom text is invalid since we don't encrypt the token but you can replace it with:

```
No because the injection of certificates to the proper place through the side-channel orchestration API. Furthermore, mutual TLS session for logging in requires the authentication client SSL key which is only kept in the authenticator client's memory and a valid/signed certificate.
```

> No. When you try and authenticate,  the Authenticator will send you an access token that’s encrypted with the private

Reviewer: ```suggestion
No. When you try and authenticate, the Authenticator will send you an access token that’s encrypted with the private
```
